core/src/main/java/com/linecorp/armeria/client/endpoint/SlowStartAwareEndpointSelectionStrategy.java

```diff
@@ -0,0 +1,175 @@
+                        .numberOfSteps(numberOfSteps)
+                        .build();
+    }
+
+    private final EndpointWeightTransition endpointWeightTransition;
+    private final ScheduledExecutorService executorService;
+    private final Duration slowStartInterval;
+    private final int numberOfSteps;
+
+    /**
+     * Creates a {@link SlowStartAwareEndpointSelectionStrategy}.
+     */
+    SlowStartAwareEndpointSelectionStrategy(EndpointWeightTransition endpointWeightTransition,
+                                            ScheduledExecutorService executorService,
+                                            Duration slowStartInterval,
+                                            int numberOfSteps) {
+        this.endpointWeightTransition = endpointWeightTransition;
+        this.executorService = executorService;
+        this.slowStartInterval = requireNonNull(slowStartInterval, "slowStartInterval");
+        checkArgument(numberOfSteps > 0, "numberOfSteps(%s) > 0");
+        this.numberOfSteps = numberOfSteps;
+    }
+
+    @Override
+    public EndpointSelector newSelector(EndpointGroup endpointGroup) {
+        return new SlowStartAwareWeightedRoundRobinSelector(endpointGroup, endpointWeightTransition);
+    }
+
+    /**
+     * An {@link EndpointSelector} that is similar to {@link WeightedRoundRobinStrategy} but it
+     * increases each {@link Endpoint} weight gradually.
+     */
+    private final class SlowStartAwareWeightedRoundRobinSelector implements EndpointSelector {
+        private final EndpointGroup endpointGroup;
+        private final AtomicInteger sequence = new AtomicInteger();
+        private volatile EndpointsAndWeights endpointsAndWeights;
+        private final EndpointWeightTransition endpointWeightTransition;
+
+        private volatile ScheduledFuture<?> updateEndpointWeightFuture;
+        private final Map<Endpoint, EndpointAndStep> currentWeights = new ConcurrentHashMap<>();
+
+        SlowStartAwareWeightedRoundRobinSelector(EndpointGroup endpointGroup,
+                                                 EndpointWeightTransition endpointWeightTransition) {
+            this.endpointGroup = endpointGroup;
+            this.endpointWeightTransition = endpointWeightTransition;
+            final List<Endpoint> endpoints = endpointGroup.endpoints();
+            endpointsAndWeights = new EndpointsAndWeights(endpoints);
+            endpointGroup.addListener(this::scheduleUpdateRequestWeight);
+        }
+
+        @Override
+        public EndpointGroup group() {
+            return endpointGroup;
+        }
+
+        @Override
+        public EndpointSelectionStrategy strategy() {
+            return SlowStartAwareEndpointSelectionStrategy.this;
+        }
+
+        @Override
+        public Endpoint select(ClientRequestContext ctx) {
+            final int currentSequence = sequence.getAndIncrement();
+            return endpointsAndWeights.selectEndpoint(currentSequence);
```

Comment:
I wonder if it's simple to mark all rampup is complete and delegate to normal endpoint selection at that point until next call to `scheduleUpdateRequestWeight`. I'd like to use this class with equal defined weights to handle rolling restart, so don't need to spend computation on weight calculation after the rampup is done.